### ISSUE TITLE
let_with_type_underscore false positive with '_ lifetime and tracing::instrument

### ISSUE THREAD

**Issue by sciguy16** [rust-lang/rust-clippy#10498]
### Summary

The new let_with_type_underscore lint is reported when an anonymous lifetime is required in a function return type when the function is instrumented with `tracing::instrument`, as in the below example. Without the `#[tracing::instrument]` clippy correctly produces no warnings. Following clippy's suggestion to remove the `'_` results in a compilation error.

### Lint Name

let_with_type_underscore

### Reproducer

I tried this code ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=615503b533515d44ae81e623ff191298)):

```rust
#[derive(Debug)]
struct Wrapper(Vec<i32>);

impl Wrapper {
    #[tracing::instrument]
    fn items(&self) -> impl Iterator<Item = i32> + '_ {
        self.0.iter().cloned()
    }
}

fn main() {
    let a = Wrapper(vec![1, 2, 3]);
    let _ = a.items();
}
```

I saw this happen:

```
warning: variable declared with type underscore
 --> src/main.rs:6:24
  |
6 |     fn items(&self) -> impl Iterator<Item = i32> + '_ {
  |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
help: remove the explicit type `_` declaration
 --> src/main.rs:6:54
  |
6 |     fn items(&self) -> impl Iterator<Item = i32> + '_ {
  |                                                      ^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#let_with_type_underscore
  = note: `#[warn(clippy::let_with_type_underscore)]` on by default
```

I expected to see this happen:
[no clippy warning]

### Version

```text
rustc 1.70.0-nightly (7b4f48927 2023-03-12)
binary: rustc
commit-hash: 7b4f48927dce585f747a58083b45ab62b9d73a53
commit-date: 2023-03-12
host: x86_64-unknown-linux-gnu
release: 1.70.0-nightly
LLVM version: 15.0.7
```

**Comment by AdamGS** [comment]
I have a similiar issue, but the code looks a bit [different](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=33c9198e1811c42eca2f1a0f160acae8) (inspired by @sciguy16's example). Not lifetimes just an `impl Iterator` return type.
```rust
#[derive(Debug)]
struct Wrapper(Vec<String>);

impl Wrapper {
    #[tracing::instrument]
    fn make_iter(self) -> impl Iterator<Item = String> {
        self.0.into_iter()
    }   
}

```

**Comment by brainstorm** [comment]
I'm seemingly hitting the same issue in https://github.com/umccr/htsget-rs/pull/171, offending code:

```rust
/// Gets the JSON to return for a service-info endpoint
#[instrument(skip(app_state))]
pub fn get_service_info_json<H: HtsGet + Send + Sync + 'static>(
  app_state: &AppState<H>,
  endpoint: Endpoint,
) -> impl Responder {
  info!(endpoint = ?endpoint, "service info request");
  PrettyJson(get_base_service_info_json(
    endpoint,
    app_state.htsget.clone(),
    &app_state.config_service_info,
  ))
}
```

And clippy error:

```rust
warning: variable declared with type underscore
  --> htsget-actix/src/handlers/service_info.rs:18:6
   |
18 | ) -> impl Responder {
   |      ^^^^^^^^^^^^^^
   |
help: remove the explicit type `_` declaration
  --> htsget-actix/src/handlers/service_info.rs:18:20
   |
18 | ) -> impl Responder {
   |                    ^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#let_with_type_underscore

warning: `htsget-actix` (lib) generated 5 warnings
    Finished dev [unoptimized + debuginfo] target(s) in 4.79s
```

## Version

```
% rustc --version
rustc 1.71.0-nightly (473f916d8 2023-05-03)
```

/cc @mmalenic

**Comment by denschub** [comment]
Jup, this seemly affects a lot of usage of `tracing`. [This Axum handler function](
https://github.com/denschub/camo-rs/blob/43f75cacb5b32758fb7e1606b72e132f8c992268/src/server.rs#L51-L57) now results in a rather confusing error from clippy:

```
warning: variable declared with type underscore
  --> src/server.rs:57:6
   |
57 | ) -> impl IntoResponse {
   |      ^^^^
   |
help: remove the explicit type `_` declaration
  --> src/server.rs:57:10
   |
57 | ) -> impl IntoResponse {
   |          ^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#let_with_type_underscore
   = note: `#[warn(clippy::let_with_type_underscore)]` on by default
```
